File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0, r = nums.size() - 1;

        while(l <= r)
        {
            int mid = (l + r) / 2;

            if(nums[mid] == target)
            {
                return mid;
            }
            else if(nums[l] <= nums[mid])
            {
                if(target < nums[mid] && target >= nums[l])
                    r = mid - 1;
                else
                    l = mid + 1;
            }
            else
            {
                if(target > nums[mid] && target <= nums[r])
                    l = mid + 1;
                else
                    r = mid - 1;
            }
        }

        return -1;
    }
};

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.cs
public class Solution {
    public int Search(int[] nums, int target) {
        int n = nums.Length;
        int l = 0, r = n - 1;
        while(l <= r)
        {
            int m = (l + r) / 2;
            if(nums[m] == target)
            {
                return m;
            }
            if(nums[l] <= nums[m])
            {
                if(target > nums[m] || target < nums[l])
                {
                    l = m + 1;
                }
                else
                    r = m - 1;
            }
            else
            {
                if(target < nums[m] || target > nums[r])
                {
                    r = m - 1;
                }
                else
                    l = m + 1;
            }
        }

        return -1;
    }
}